ENH: interpolate.BSpline: add array API standard interface

ev-br opened the pull request:
This PR adds a minimal Array API interface to `BSpline` and `make_interp_spline`.

There are several minor details worth spelling out. First, `BSpline` uses a C extension, therefore non-numpy array inputs need to go through the now-standard `np.asarray(input); result = c_function(input); return xp.asarray(result)` dance.

Second, `BSpline` instances used to store numpy arrays. Namely, for array-like `t` and `c`, and  `spl = BSpline(t, c, k)`, then `spl.t` and `spl.c` are numpy array attributes derived from the `__init__` arguments.
This PR changes these attributes into properties, so that if `t` and `c` are torch tensors, and `spl = BSpline(t, c, k)`, 
a new private attribute `spl._t` is a numpy array, and `spl.t` is a propery which constructs a torch tensor from `spl._t`.

The advantage of this is that at the call time, we can use `spl._t` without paying the price of `np.asarray` (which is not negligible here, actually). An unpleasant consequence is that the property abstraction leaks for mutable properties: `spl.t.fill(0)` goes through the getter not the setter. There's nothing to be done about it though, and this does not lead to new failure modes AFAICS.

Third, if `t`, `c` and `x` are torch tensors, then `spl = BSpline(t, c, k); spl(x)` should produce a torch tensor, too. To achieve this, we store the namespace on an instance, via `self.xp = array_namespace(t, c)` and the `__call__` is

```
def __call__(self, x):
     x_np = np.asarray(x)
     out = heavy_lifting_in_c(x_np)
     return self.xp.asarray(out)
```

At the moment, the new attribute `xp` is public, since I don't see a particularly strong need to hide it. If anything, it can be useful to a user: otherwise there's no easy way for a user to tell what is the type of `spl(1)` without calling it.
If this turns out to be problematic, we can always hide the attribute behind a property. Or store `xp.asarray` as a private method. 

One last small thing: since module objects are not picklable, we need to add a custom pickler to correctly handle the `xp` attribute.

ev-br committed:
ENH: interpolate: a minimal array API implementation for BSpline etc

convert BSpline (minus design_matrix, which returns a csr_array),
splder, splantider, make_interp_spline

Since heavy lifting is in C, only do array_namespace on inputs,
convert to numpy, do the work, convert to self.xp on exit.

ev-br committed:
TST: interpolate: test BSpline array API

ev-br committed:
TST: interpolate: skip CPU-only tests

rgommers commented on scipy/interpolate/_fitpack_impl.py:
```diff
@@ -745,13 +748,23 @@ def insert(x, tck, m=1, per=0):
 
-def splder(tck, n=1):
+# np.r_ replacement
+def npr(xp, *arys):
```
A readable name here please! I suggest `concat_1d`.

rgommers replied:
With a code comment or docstring explaining that this exists because `xp.concat` doesn't take scalars or 0-D arrays.

lucascolley replied:
I think we should also consider putting helpers like this in `_lib._array_api`. I haven't blocked on this when merging previous PRs, but it does seem a bit messy to have such general functions living in specific submodules. It also would make it easier to see at a glance the helpers which we have written.

rgommers reviewed (commented):
Overall LGTM, thanks Evgeni.

> > At the moment, the new attribute `xp` is public, since I don't see a particularly strong need to hide it.

Private please! There is no need to leak this, even after array API support becomes fully public.

> If anything, it can be useful to a user: otherwise there's no easy way for a user to tell what is the type of `spl(1)` without calling it. 

If this info is needed - which should be rare - `isinstance` on one of the array attributes is still quite cheap, and more idiomatic.

> If this turns out to be problematic, we can always hide the attribute behind a property. Or store `xp.asarray` as a private method.

Renaming it from `.xp` to `._xp` seems most straightforward here.

ev-br committed:
MAINT: interpolate/bspline: self._xp not self.xp

ev-br committed:
MAINT: _lib:  add concat_1d, a replacement for np.r_

xp.concat does not accept python scalars or 0D arrays,
so we need a helper for thing like `np.r_[0, x]` and
`np.r_[x[0], x]`.

ev-br committed:
MAINT: signal: use concat_1d

ev-br committed:
MAINT: interpolate: cater for JAX in BSpline methods/tests

ev-br commented:
Comments addressed. 

Moving `concat_1d` into `_lib` runs into some sort of  import cycle between array-api-compat and array-api-extra. Using a local import fixes the immediate problem, but I suppose it is brittle either way. The multivendoring contraption there is frankly rather crazy IMO. Bringing it back to sanity is out of scope for this PR, so if we'd rather avoid local imports, I can move concat_1d back to scipy.interpolate.

lucascolley commented:
Good point. I think the simplest fix may be to move `_array_api.array_namespace` to a different file which doesn't import from `_array_api`.

ev-br committed:
MAINT: interpolate: remove the need for a custom pickler

Store a private `_asarray` function on a BSpline instance instead
of the namespace. This way, we do not need a custom getstate/setstate
or reduce or any other custom pickle support.

ev-br commented:
>One last small thing: since module objects are not picklable, we need to add a custom pickler to correctly handle the xp attribute.

Actually, we don't if we store a private function instead of a private namespace: instead of `self._xp.asarray` we store `self._asarray = array_namespace(t, c).asarray` and all is well. The need to call other namespaced functions should be rare, and here it only happens in `derivative` and `antiderivative`, which are essentially constructors. So no performance considerations here, and the code is simpler, I'd say it's worth it.
The last commit makes this change.

j-bowhay reviewed (commented):
Thanks, not looked at the tests yet

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -328,11 +349,15 @@ def basis_element(cls, t, extrapolate=True):
         c = np.zeros_like(t)
         c[k] = 1.
+
+        t, c = xp.asarray(t), xp.asarray(c)
```
Bugs me too, a bit. This method is not meant to be in a tight loop though, and the parts that are, have numpy arrays to work on.

lorentzenchr commented on scipy/_lib/_array_api.py:
```diff
@@ -651,6 +653,15 @@ def xp_result_device(*args):
 
+# np.r_ replacement
+def concat_1d(xp, *arys):
+    """A replacement for np.r_: `xp.concat` does not accept python scalars or 0D arrays.
```
```suggestion
    """A replacement for `np.r_` as `xp.concat` does not accept python scalars or 0-D arrays.
```

lorentzenchr commented on scipy/_lib/_array_api.py:
```diff
@@ -651,6 +653,15 @@ def xp_result_device(*args):
 
 
+# np.r_ replacement
+def concat_1d(xp, *arys):
```
```suggestion
def concat_1d(xp: ModuleType | None, *arrays: Iterable[ArrayLike]) -> Array:
```
In line with other functions in this file.

lorentzenchr commented on scipy/interpolate/_bsplines.py:
```diff
@@ -271,6 +276,22 @@ def tck(self):
         """
         return self.t, self.c, self.k
 
+    @property
```
A comment here or above in the init would help, something like
```suggestion
    # Under the hood, self._c and self._t are always saved as numpy array
    # because they are used in a C extension expecting numpy arrays.
    @property
```

lorentzenchr reviewed (approved):
What happens if c and t are different arrays like `BSpline(t=torch_array, c=numpy_array, 2)`? (Not that it makes sense, but in case a user does it.)

ev-br committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

ev-br commented:
> What happens if c and t are different arrays like BSpline(t=torch_array, c=numpy_array, 2)? (Not that it makes sense, but in case a user does it.)

What Lucas said: mixing array types in the constructor is an automatic TypeError from `array_namespace`. What is somewhat less clear-cut is mixing array types between `__init__` and `__call__` (in the scikit-learn language: `.fit` and `.predict`):  what is `BSpline(t=np_array, c=np_array, k)(torch_tensor)`?

In this PR --- and I think this is correct in general --- the logic is that the array namespace is fixed by the arguments of the constructor; at call time, the arguments are cast to the array type _of that fixed namespace_:

```
In [1]: import torch

In [2]: import numpy as np

In [3]: from scipy.interpolate import BSpline, make_interp_spline

In [4]: spl = make_interp_spline(np.arange(10), np.arange(10)**2)

In [5]: spl_ = BSpline(torch.asarray(spl.t), torch.asarray(spl.c), spl.k)

In [7]: spl_(1)
Out[7]: tensor(1., dtype=torch.float64)

In [8]: spl_(np.asarray(1))
Out[8]: tensor(1., dtype=torch.float64)
```

ev-br commented on scipy/_lib/_array_api.py:
```diff
@@ -651,6 +653,15 @@ def xp_result_device(*args):
+    """A replacement for np.r_: `xp.concat` does not accept python scalars or 0D arrays.
+    """
+    arys = [xp.asarray(a) for a in arys]
+    arys = [xpx.atleast_nd(a, ndim=1, xp=xp) for a in arys]
```
Included, thanks.

ev-br commented:
Addressed all review comments.

lorentzenchr commented:
> In this PR --- and I think this is correct in general --- the logic is that the array namespace is fixed by the arguments of the constructor; at call time, the arguments are cast to the array type of that fixed namespace

Maybe it does not matter much and consistency within scipy is more import, but as a user I would expect to get the same array back as I pass into `__call__`.

lucascolley reviewed (commented):
thanks Evgeni, very close!

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -331,11 +354,15 @@ def basis_element(cls, t, extrapolate=True):
 
         """
-        k = len(t) - 2
+        xp = array_namespace(t)
+        t = np.asarray(t)
+        k = t.shape[0] - 2
         t = _as_float_array(t)
```
No opinion on renaming `_as_float_array`.  Send a patch if you have feel strongly about it?

----- 

Conversions to/from numpy: there are of course multiple ways to skin this one. It's used ten times, 
```
$ git grep -n "_as_float_array" scipy/interpolate/ | wc -l
11
 ```
and IIRC other uses are after the conversion to numpy. So it felt easier to just do it like it's done here and be done with it. Care to send a patch if you see a way to make it nicer across the module?

lucascolley replied:
Just to clarify, the only reason we are moving to `np` here is to use `_as_float_array`? If so, that seems bad.

lucascolley replied:
alright, if you add a TODO comment here to avoid the namespace conversion I am happy to merge!

Alternatively, I could try to come up with a patch in the coming days.

ev-br replied:
A patch would be great indeed @lucascolley! Meanwhile I'll then add a TODO comment to gh-23052 to avoid flushing the CI here.

lucascolley reviewed (approved):
thanks Evgeni, and for the review @lorentzenchr !
